kratos/tensor_adaptors/tensor_adaptor.h

```diff
@@ -0,0 +1,421 @@
+ *                  model_part.CreateNewNode(i + 1, 0.0, 0.0, 0.0)->SetValue(PRESSURE, i + 1);
+ *              }
+ *
+ *              VariableTensorAdaptor var_ta(model_part.pNodes(), PRESSURE);
+ *              var_ta.Check();       // Checks if the PRESSURE variable is available in each of the nodes.
+ *              var_ta.CollectData(); // Reads the PRESSURE variable from each of the nodes and puts them into an internal storage.
+ *
+ *              auto data = var_ta.ViewData(); // Create a span viewing the data
+ *
+ *              // here we modify the internal data storage of the TensorAdaptor.
+ *              // This will not modify the PRESSURE of the nodes.
+ *              for (IndexType i = 0; i < data.size(); ++i) {
+ *                  data[i] += 1.0;
+ *              }
+ *
+ *              var_ta.StoreData(); // Writes back the modified tensor adaptor data back to each node's PRESSURE variable.
+ *          @endcode
+ *          The following is the equivalent python example.
+ *          @code{.py}
+ *              import KratosMultiphysics as Kratos
+ *
+ *              model = Kratos.Model()
+ *              model_part = model.CreateModelPart("test")
+ *              for i in range(10):
+ *                  model_part.CreateNewNode(i + 1, 0.0, 0.0, 0.0).SetValue(Kratos.PRESSURE, i + 1)
+ *
+ *              var_ta = Kratos.TensorAdaptors.VariableTensorAdaptor(model_part.Nodes(), Kratos.PRESSURE)
+ *              var_ta.Check()
+ *              var_ta.CollectData()
+ *              var_ta.data += 1.0
+ *              var_ta.StoreData()
+ *          @endcode
+ */
+
+ /**
+ * @class TensorAdaptor
+ * @ingroup TensorAdaptors
+ *
+ * @section supported_container Supported container types
+ * - @ref ModelPart::DofsArrayType
+ * - @ref ModelPart::NodesContainerType
+ * - @ref ModelPart::ConditionsContainerType
+ * - @ref ModelPart::ElementsContainerType
+ * - @ref ModelPart::PropertiesContainerType
+ * - @ref ModelPart::GeometryContainerType
+ * - @ref ModelPart::MasterSlaveConstraintContainerType
+ *
+ * @author Suneth Warnakulasuriya
+ * @brief Provides an abstraction for managing tensor data associated with Kratos ModelPart containers.
+ * @tparam TDataType The type of the data stored in the tensor adaptor.
+ */
+template<class TDataType>
+class KRATOS_API(KRATOS_CORE) TensorAdaptor {
+protected:
+    ///@name Class definitions
+    ///@{
+
+    /**
+     * @class Storage
+     * @brief Manages the storage and lifetime of tensor adaptor data associated with various Kratos ModelPart containers.
+     *
+     * The Storage class encapsulates a pointer to a container (such as nodes, elements, conditions, etc.) and manages
+     * the associated tensor data. It provides mechanisms for copying, moving, and viewing the internal data, as well as
+     * querying the shape and size of the tensor. Reference counting is implemented for safe memory management using
+     * intrusive pointers.
```

Comment:
*sigh*. I can't say I'm enthusiastic about AI docs.

```suggestion
     * querying the shape and size of the tensor.
```